### PR TITLE
Add bitwarden_rs helm-chart

### PR DESCRIPTION
**Pull request by Ornias1993**
Due to much community request:
A Helm-Chart (based on the work from GissiLabs), that includes all possible options for bitwarden_rs.

To be sure it doesn't need too much babysitting: it has it's own CI already setup and ready to go.
CI tests both pushes and pullrequests. It does both a Lint test and actual install test.

Releases for the charts would be on every commit pushed to master, with [release chart] in the title.
(this prevents accidental releases overwriting the current version and such.)

Future TODO's:
- Add a chart-version check to the CI and update if chart version is updated

**Requirements:**
Needs github pages to be setup and linked to a branch called gh-pages. (before being merged)
No actual website will be published, just a webspot to serve the packaged helmcharts.

**TLDR:**
- Adds a "contrib" directory for community contributions (that are not directly integrated into bitwarden_rs)
- Adds a helm-chart for the project
- Adds Test and Releasing CI for the helm chart
- Some cleanup

Co-authored-by: gissilabs
Signed-off-by: Kjeld Schouten-Lebbing <kjeld@schouten-lebbing.nl>

### CODE REVIEW

**Review by BlackDex — COMMENTED** [review]
Looks interesting, have to look a bit better at it, but one thing that i noticed is that there is the apache license, while i think we are using GPLv3.

**Comment by Ornias1993** [comment]
There is no code being linked, hence virality does not apply and the original was 
apache. So it is apache.

**Comment by Ornias1993** [comment]
The thing about “contrib” code (scripts, charts etc), is that it does not have to follow the project license: contributed additions.

**Comment by Ornias1993** [comment]
Apache is GPL compatible btw, according tot the fsf.

**Comment by Ornias1993** [comment]
This PR is based on works that have been originaly opensourced as Apache.
That license file is required.

In accordance with the FSF and SFLC advisory about including apache licensed works in GPLv3 projects, this does not create any problem. More so because this is not linked with the GPLv3 work, its just a contrib(uted) install script.

**Comment by Ornias1993** [comment]
update:
Edited the readme to clearify the license of the charts.

**Comment by Ornias1993** [comment]
update:
- Fixed co-authored-by tags 
- Add all upstream authors as co-authors
- Clearly refered to upstream source in commit

**Comment by dani-garcia** [comment]
Is there any reason this benefits from being in this repo instead of a third party one that we can link in the wiki? Maybe these changes can be submitted to the already existing helm chart by Skeen https://github.com/dani-garcia/bitwarden_rs/wiki/Deployment-examples

So far I've tried to keep away from handling any other packaging/deployment system apart from Docker because there's a lot of them and I don't have the time nor the knowledge to maintain them.

**Comment by Ornias1993** [comment]
@dani-garcia 
> Is there any reason this benefits from being in this repo instead of a third party one that we can link in the wiki?

- Having a official and single repo for everyone to contribute to. 
- If you want to support docker: Kubernetes and Helm are the leading deployment systems for docker containers. 

TLDR: it would be nice to officially host a chart for people to contribute to instead of dozens of current charts of variable quality.

> Maybe these changes can be submitted to the already existing helm chart by Skeen 

The one from Skeen is actually not one of the best charts i've seen. Not offense intended to Skeen, but you want me to completely overhaul Skeen's chart, others ask for it to be included with k8s@home's chart and so forth. Thats why I think it's time to have one chart for the community to work on.

> So far I've tried to keep away from handling any other packaging/deployment system

Helm = docker, it's a way to get your docker container into k8s. Not a new package or deployment. It does NOTHING with bitwardenrs itself, it just points helm to the docker container and specifies default settings to be used for the container.

> I don't have the time nor the knowledge to maintain them.

- Thats why I started the /contrib directory, just like many other projects those are community additions (most often deployment scripts or external build tools) that are not officially maintained by the project (but unofficially often are)
- Thats also why I included a complete CI suite, to at least validate PR's for basic quality.
- If you want, have a talk with the upstream developer, he is willing to assist in QA and maintenance it seems.

**Comment by sgissi** [comment]
Hi, I'm the maintainer of the chart. My comments have been added on the Dual-license PR (https://github.com/gissilabs/charts/pull/14). I'm not particularly found of the idea of a 3rd party submitting my work without consent but I do stand by my comment: If the Bitwarden_rs team is willing to take over maintenance of the chart officially, I'm more than happy to have it in this repo and contribute when possible. If it is to be a community-managed chart, I rather keep it in my own repository. I usually update to the latest image a couple weeks after release just to avoid exposing early bugs. I do maintain it on my spare time so delays might happen :)

As for maintenance burden, it is not a lot of work: new options need to be updated on a couple YAML and README files, some logic might be needed if there are conflicting options or dependencies.

**Comment by Ornias1993** [comment]
> I'm not particularly found of the idea of a 3rd party submitting my work without consent

In opensource there are already too many people mixing their feelings about porting and forking with opensource. Please keep feelings out of it. Code is either good enough to be ported (and maintained) or it's not. You are free to have them, but voicing them is pushing this into a direction that isn't code related at all, while it should be.

>  If it is to be a community-managed chart

It will be community maintained, the goal is to have a central chart for the community (including you) to contribute to. Not for someone else (including me) to take over. Although I'm willing to port any changes you make.

> I usually update to the latest image a couple weeks after release just to avoid exposing early bugs. I do maintain it on my spare time so delays might happen :)

I think you have a great trackrecord on that indeed! :)

> As for maintenance burden, it is not a lot of work: new options need to be updated on a couple YAML and README files, some logic might be needed if there are conflicting options or dependencies.

I think last weeks update is a good example:
a few rows in the readme with new features and a few in the templates.

It also has to be noted that almost any minor or patch update of bitwardenrs should stay mostly compatible, the current env-vars almost never change significantly. Adding new features to the chart is not technically required for it to stay working.

TLDR: Indeed it's pretty hands-off when it comes to maintenance.

**Comment by CuBiC3D** [comment]
>others ask for it to be included with k8s@home's chart and so forth.

I was the one asking this on reddit, since one already exists in a well known community organisation.

But I completely agree, the chart in this PR seems to be very well made and maintained and I would love to see this in the official repository. Helm and Kubernetes are de facto standards for container deployment and orchestration. Especially to have a trusted source for a chart is important in the context of bitwarden.

**Comment by Ornias1993** [comment]
> I was the one asking this on reddit, since one already exists in a well known community organisation.

Uhmm, I didn't know it and also don't know who's behind it. Not saying it's bad, but "well known" is very dependend on the eye of the beholder ;-)

 

> Helm and Kubernetes are de facto standards for container deployment and orchestration. Especially to have a trusted source for a chart is important in the context of bitwarden.

Yes, thats also a big thing: Trust.
If something is official we can at least expect the wider Bitwarden_RS community had the chance to look at PR's and such.

**Comment by BlackDex** [comment]
I agree with @dani-garcia regarding the staying away from third-party packaging and building stuff. Besides from having Dockerfiles so that this project can built docker images via docker hub.

A separate wiki page or something added to an existing wiki page regarding where to get information about using Helm would be enough for me.

Maybe we could update the README.md with links to multiple options so that people can find this more quickly, but besides that, i think we should keep stuff like this upstream with people who use this instead of have all these third-party items added to this repo.

So i am in favor of closing this PR because of the above.

**Comment by Ornias1993** [comment]
@BlackDex Helm is not packaging nor building, it's just install scripting. Which I think is significantly different.
Bitwarden_RS is barely useable with K8S without such a system or at least a very significant burden.

Considering all self-respecting projects are slowly also including a form of K8S support, I think it's only normal to include.

"have all these third-party items added to this repo."
Thats not what i'm asking, i'm not asking for "all these" things, i'm asking for an industry standard to be supported. "Don't use it /care" seems arogant at best.

**Comment by Ornias1993** [comment]
Anyhow: If no maintainer wants to run a significant community behind this, might as wel close. Because it isn't going to get merged anyway.

**Comment by BlackDex** [comment]
I'm sorry to see that it seems there is no room for discussion about this. Thanks for sharing your point of view at least.
